### PR TITLE
security: Fix path traversal vulnerability in SaveVisualizationAsFile

### PR DESCRIPTION
**Pull request by Dor-bl**
🎯 **What:** The `SaveVisualizationAsFile` method lacked validation on the user-supplied `fileName`, directly writing the Base64 visualization to the provided path.
⚠️ **Risk:** An attacker controlling the filename could provide an absolute path (`C:\Windows\System32\malicious.dll`) or use relative path traversal (`../../etc/passwd`) to overwrite arbitrary files on the local filesystem.
🛡️ **Solution:** The method now restricts the output file location to the application's current working directory. It uses `Path.GetFullPath` combined with `StartsWith` (aware of target operating system case sensitivity) to ensure that no directory traversal sequences or absolute paths escape the allowed base directory. Symbolic links/reparse points are also detected and blocked. Invalid path characters are rejected by throwing an exception.

### CODE REVIEW

File: src/Appium.Net/Appium/ImageComparison/ComparisonResult.cs [hunk 1]
```diff
@@ -21,7 +22,92 @@ protected ComparisonResult(Dictionary<string, object> result)
...
+                !allowedDirectory.EndsWith(Path.AltDirectorySeparatorChar.ToString()))
+            {
+                allowedDirectory += Path.DirectorySeparatorChar;
+            }
+
+            string fullPath = Path.GetFullPath(Path.Combine(currentDirectory, fileName));
+
+            var comparison = (RuntimeInformation.IsOSPlatform(OSPlatform.Windows) || RuntimeInformation.IsOSPlatform(OSPlatform.OSX))
+                ? StringComparison.OrdinalIgnoreCase
+                : StringComparison.Ordinal;
```
[the remark this answers is not in the document]
Comment (Dor-bl) [reply src/Appium.Net/Appium/ImageComparison/ComparisonResult.cs]: Done. The code has been updated to only use `OrdinalIgnoreCase` on Windows and `Ordinal` on macOS and Linux (so case-sensitive sibling directory traversals will be blocked correctly on macOS/Linux).

File: src/Appium.Net/Appium/ImageComparison/ComparisonResult.cs [hunk 2]
```diff
@@ -21,7 +22,92 @@ protected ComparisonResult(Dictionary<string, object> result)
...
+                        // If attributes cannot be read due to permissions, treat the path as unsafe.
+                        return true;
+                    }
+
+                    if ((attributes & FileAttributes.ReparsePoint) != 0)
+                    {
+                        // A reparse point (including symlinks/junctions) is present along the path.
+                        return true;
+                    }
+                }
```
[the remark this answers is not in the document]
Comment (Dor-bl) [reply src/Appium.Net/Appium/ImageComparison/ComparisonResult.cs]: Done. The symlink detection has been rewritten to check if the path is a symbolic link or reparse point even if the target does not exist. We query the `.NET 6+` `LinkTarget` property on `FileSystemInfo` via reflection at runtime. If reflection is not available (e.g. on `.NET Framework 4.8` on Windows), we verify if the entry exists in its parent directory using `Directory.GetFileSystemEntries` and check the reparse point flag.

File: test/integration/ImageComparison/ComparisonResultTests.cs [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+                Assert.Ignore("Skipping symlink test: Symbolic links are not supported on this target framework.");
+                return;
+#endif
+
+                // Attempt to write a file via the symlink
+                string fileName = Path.Combine("local_sub", "symlink_dir", "image.png");
+
+                var ex = Assert.Throws<IOException>(() => _comparisonResult.SaveVisualizationAsFile(fileName));
+                Assert.That(ex.Message, Does.Contain("symbolic link or reparse point"));
+            }
```
[the remark this answers is not in the document]
Comment (Dor-bl) [reply test/integration/ImageComparison/ComparisonResultTests.cs:189]: Done. Added the test case `SaveVisualizationAsFile_BrokenFileSymlinkTraversal_ThrowsIOException` to cover file symlinks whose targets do not exist yet.

File: src/Appium.Net/Appium/ImageComparison/ComparisonResult.cs [hunk 4]
```diff
@@ -21,7 +25,140 @@ protected ComparisonResult(Dictionary<string, object> result)
...
+
+            if (fileName.Length == 0)
+            {
+                throw new ArgumentException("The file name must not be an empty string.", nameof(fileName));
+            }
+
+            if (fileName.IndexOfAny(Path.GetInvalidPathChars()) >= 0)
+            {
+                throw new ArgumentException("The file name contains invalid characters.", nameof(fileName));
+            }
```
[the remark this answers is not in the document]
Comment (Dor-bl) [reply src/Appium.Net/Appium/ImageComparison/ComparisonResult.cs]: Done. Added an explicit check on Windows that rejects any colons in the filename unless they are part of a drive letter specifier at the start of the path (e.g. `C:\\`), protecting against alternate data stream writes.

File: test/integration/ImageComparison/ComparisonResultTests.cs [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+        private const string DummyBase64 = "c29tZSB2aXN1YWxpemF0aW9uIGRhdGE="; // "some visualization data"
+        private byte[] _expectedBytes;
+
+        [SetUp]
+        public void SetUp()
+        {
+            _testDir = Path.Combine(Directory.GetCurrentDirectory(), "test_output_" + Guid.NewGuid().ToString("N"));
+            Directory.CreateDirectory(_testDir);
+            Directory.SetCurrentDirectory(_testDir);
+
```
[the remark this answers is not in the document]
Comment (Dor-bl) [reply test/integration/ImageComparison/ComparisonResultTests.cs:33]: Done. Stored the original working directory in `_originalCwd` during SetUp and restored it during TearDown.

File: test/integration/ImageComparison/ComparisonResultTests.cs [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+            _comparisonResult = new TestComparisonResult(dict);
+            _expectedBytes = Convert.FromBase64String(DummyBase64);
+        }
+
+        [TearDown]
+        public void TearDown()
+        {
+            // Restore current directory before deleting
+            string originalDir = Path.GetFullPath(Path.Combine(_testDir, ".."));
+            Directory.SetCurrentDirectory(originalDir);
```
[the remark this answers is not in the document]
Comment (Dor-bl) [reply test/integration/ImageComparison/ComparisonResultTests.cs]: Done. Restored `_originalCwd` directly in TearDown.

File: src/Appium.Net/Appium/ImageComparison/ComparisonResult.cs [hunk 7]
```diff
@@ -21,7 +25,157 @@ protected ComparisonResult(Dictionary<string, object> result)
...
+            string allowedDirectory = Path.GetFullPath(currentDirectory);
+            if (!allowedDirectory.EndsWith(Path.DirectorySeparatorChar.ToString()) &&
+                !allowedDirectory.EndsWith(Path.AltDirectorySeparatorChar.ToString()))
+            {
+                allowedDirectory += Path.DirectorySeparatorChar;
+            }
+
+            string fullPath = Path.GetFullPath(Path.Combine(currentDirectory, fileName));
+
+            var comparison = RuntimeInformation.IsOSPlatform(OSPlatform.Windows)
```
Comment (KazuCocoa) [inline src/Appium.Net/Appium/ImageComparison/ComparisonResult.cs]: We have multiple `RuntimeInformation.IsOSPlatform`, so they could be reduce the number of call

**Review by KazuCocoa — APPROVED** [review]
Left minor comment.

I don't recall if we have similar code in this repo to accept path as string, but we could extract this path check as a "isValidPath", for example, and we can call the method as well

File: src/Appium.Net/Appium/ImageComparison/ComparisonResult.cs (hunk 7 above)
Comment (Dor-bl) [reply src/Appium.Net/Appium/ImageComparison/ComparisonResult.cs]: Done. Extracted the path validation into `PathHelpers.ValidateAndGetFullPath` and cached `IsWindows` to reduce runtime overhead.

File: src/Appium.Net/Appium/ImageComparison/ComparisonResult.cs [hunk 8]
```diff
@@ -1,7 +1,8 @@
-﻿using System;
+using System;
 using System.Collections.Generic;
 using System.Drawing;
 using System.IO;
+using System.Runtime.InteropServices;
```
[the remark this answers is not in the document]
Comment (Dor-bl) [reply src/Appium.Net/Appium/ImageComparison/ComparisonResult.cs]: It's automatically resolved by C# because `OpenQA.Selenium.Appium.ImageComparison` is a sub-namespace of `OpenQA.Selenium.Appium`. Regardless, I've removed the unused `using System.Runtime.InteropServices;`.

File: src/Appium.Net/Appium/PathHelpers.cs [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+                }
+
+                if (!Directory.Exists(parent))
+                {
+                    return false;
+                }
+
+                string name = Path.GetFileName(path);
+                string[] entries = Directory.GetFileSystemEntries(parent, name);
+                return entries != null && entries.Length > 0;
```
[the remark this answers is not in the document]
Comment (Dor-bl) [reply src/Appium.Net/Appium/PathHelpers.cs]: Great catch! Fixed to enumerate all entries and check for an exact match.

File: src/Appium.Net/Appium/PathHelpers.cs [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+
+            if (fileName.Length == 0)
+            {
+                throw new ArgumentException("The file name must not be an empty string.", nameof(fileName));
+            }
+
+            if (fileName.IndexOfAny(Path.GetInvalidPathChars()) >= 0)
+            {
+                throw new ArgumentException("The file name contains invalid characters.", nameof(fileName));
+            }
```
[the remark this answers is not in the document]
Comment (Dor-bl) [reply src/Appium.Net/Appium/PathHelpers.cs:37]: Done. I've updated the logic to validate each segment using `Path.GetInvalidFileNameChars()`, which correctly catches `*`, `?`, and others on Windows, while preserving the explicit allowlist logic for the Windows drive specifier `:`.

File: src/Appium.Net/Appium/PathHelpers.cs [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+                string name = Path.GetFileName(path);
+                var comparison = IsWindows ? StringComparison.OrdinalIgnoreCase : StringComparison.Ordinal;
+                string[] entries = Directory.GetFileSystemEntries(parent);
+                foreach (string entry in entries)
+                {
+                    if (Path.GetFileName(entry).Equals(name, comparison))
+                    {
+                        return true;
+                    }
+                }
```
[the remark this answers is not in the document]
Comment (Dor-bl) [reply src/Appium.Net/Appium/PathHelpers.cs]: Good point! Switched to `Directory.EnumerateFileSystemEntries`.

File: src/Appium.Net/Appium/ImageComparison/ComparisonResult.cs [hunk 12]
```diff
@@ -21,7 +21,8 @@ protected ComparisonResult(Dictionary<string, object> result)
 
         public void SaveVisualizationAsFile(string fileName)
         {
-            File.WriteAllBytes(fileName, Convert.FromBase64String(Visualization));
+            string fullPath = PathHelpers.ValidateAndGetFullPath(fileName);
+            File.WriteAllBytes(fullPath, Convert.FromBase64String(Visualization));
         }
```
[the remark this answers is not in the document]
Comment (Dor-bl) [reply src/Appium.Net/Appium/ImageComparison/ComparisonResult.cs:27]: You're absolutely right about the TOCTOU risk. Unfortunately, .NET does not expose cross-platform `O_NOFOLLOW` semantics in `FileStream` or `File.WriteAllBytes`. Given the scope of this PR, the current validation significantly raises the bar against path traversal by catching structural attempts, but fully mitigating TOCTOU would require complex P/Invokes (e.g., `CreateFile` with `FILE_FLAG_OPEN_REPARSE_POINT` on Windows, `open` with `O_NOFOLLOW` on POSIX). We can track this as a separate issue if full TOCTOU mitigation is required.

File: src/Appium.Net/Appium/ImageComparison/ComparisonResult.cs [hunk 13]
```diff
@@ -21,7 +21,8 @@ protected ComparisonResult(Dictionary<string, object> result)
 
         public void SaveVisualizationAsFile(string fileName)
         {
-            File.WriteAllBytes(fileName, Convert.FromBase64String(Visualization));
+            string fullPath = PathHelpers.ValidateAndGetFullPath(fileName);
```
[the remark this answers is not in the document]
Comment (Dor-bl) [reply src/Appium.Net/Appium/ImageComparison/ComparisonResult.cs:25]: In C#, declaring `namespace A.B` means that all types in namespace `A` are implicitly in scope. I have verified that this builds successfully on both .NET 8.0 and .NET Framework 4.8.

File: test/integration/ImageComparison/ComparisonResultTests.cs [hunk 14]
```diff
@@ -0,0 +1,232 @@
...
+        public void SaveVisualizationAsFile_InvalidCharacters_ThrowsArgumentException()
+        {
+            char[] invalidChars = Path.GetInvalidPathChars();
+            if (invalidChars.Length > 0)
+            {
+                string invalidFileName = "test" + invalidChars[0] + ".png";
+                var ex = Assert.Throws<ArgumentException>(() => _comparisonResult.SaveVisualizationAsFile(invalidFileName));
+                Assert.That(ex.Message, Does.Contain("The file name contains invalid characters."));
+            }
+        }
```
[the remark this answers is not in the document]
Comment (Dor-bl) [reply test/integration/ImageComparison/ComparisonResultTests.cs]: Done. I have replaced the silent pass with an explicit `Assert.Ignore` for platforms that do not define any invalid characters, and switched to using `Path.GetInvalidFileNameChars()` for consistency with the main implementation.

File: src/Appium.Net/Appium/ImageComparison/ComparisonResult.cs [hunk 15]
```diff
@@ -21,7 +21,8 @@ protected ComparisonResult(Dictionary<string, object> result)
...
+            File.WriteAllBytes(fullPath, Convert.FromBase64String(Visualization));
```
[the remark this answers is not in the document]
Comment (Dor-bl) [reply src/Appium.Net/Appium/ImageComparison/ComparisonResult.cs:26]: Even though it compiles fine as-is because C# automatically searches parent namespaces, I've gone ahead and added the explicit `using OpenQA.Selenium.Appium;` to appease the analyzer.

File: src/Appium.Net/Appium/PathHelpers.cs [hunk 16]
```diff
@@ -0,0 +1,180 @@
...
+            }
+
+            string currentDirectory = Directory.GetCurrentDirectory();
+            string allowedDirectory = Path.GetFullPath(currentDirectory);
+            if (!allowedDirectory.EndsWith(Path.DirectorySeparatorChar.ToString()) &&
+                !allowedDirectory.EndsWith(Path.AltDirectorySeparatorChar.ToString()))
+            {
+                allowedDirectory += Path.DirectorySeparatorChar;
+            }
```
[the remark this answers is not in the document]
Comment (Dor-bl) [reply src/Appium.Net/Appium/PathHelpers.cs:75]: The goal of this validation is to prevent the provided `fileName` (untrusted input) from escaping the base directory (the current working directory). If the application intentionally starts the process in a symlinked working directory, writing within that intended directory is expected behavior and not a path traversal vulnerability in this library.

File: test/integration/ImageComparison/ComparisonResultTests.cs [hunk 17]
```diff
@@ -0,0 +1,234 @@
...
+        [Test]
+        public void SaveVisualizationAsFile_InvalidCharacters_ThrowsArgumentException()
+        {
+            char[] invalidChars = Path.GetInvalidFileNameChars();
+            if (invalidChars.Length == 0)
+            {
+                Assert.Ignore("No invalid characters defined on this platform.");
+            }
+
+            string invalidFileName = "test" + invalidChars[0] + ".png";
```
[the remark this answers is not in the document]
Comment (Dor-bl) [reply test/integration/ImageComparison/ComparisonResultTests.cs]: Excellent point! Since `ValidateAndGetFullPath` treats directory separators as valid for relative paths, using a separator as the invalid character in the test could inadvertently produce a valid subdirectory path instead of an invalid filename. I've updated the test to explicitly exclude directory separator characters from the list of invalid characters before testing.

File: src/Appium.Net/Appium/PathHelpers.cs [hunk 18]
```diff
@@ -0,0 +1,180 @@
...
+                    if (Array.IndexOf(invalidFileNameChars, c) >= 0)
+                    {
+                        if (IsWindows && c == ':' && i == 0 && part.Length == 2 && char.IsLetter(part[0]) && part[1] == ':')
+                        {
+                            continue;
+                        }
+                        throw new ArgumentException("The file name contains invalid characters or alternate data streams.", nameof(fileName));
+                    }
+                }
+            }
```
[the remark this answers is not in the document]
Comment (Dor-bl) [reply src/Appium.Net/Appium/PathHelpers.cs:67]: Good catch. While standard device names without extensions (like `CON`) might be rejected by the base-directory check if `Path.GetFullPath` converts them to absolute paths like `\\.\CON`, it's safer and more robust to explicitly reject them. I've added a check that rejects all Windows reserved device names in any path segment, regardless of their extension.

File: src/Appium.Net/Appium/PathHelpers.cs [hunk 19]
```diff
@@ -0,0 +1,198 @@
...
+                "LPT0", "LPT1", "LPT2", "LPT3", "LPT4", "LPT5", "LPT6", "LPT7", "LPT8", "LPT9"
+            };
+
+            for (int i = 0; i < parts.Length; i++)
+            {
+                string part = parts[i];
+
+                if (IsWindows)
+                {
+                    string baseName = Path.GetFileNameWithoutExtension(part);
```
[the remark this answers is not in the document]
Comment (Dor-bl) [reply src/Appium.Net/Appium/PathHelpers.cs]: Excellent point. Win32 path normalization does strip trailing spaces and dots, meaning `CON ` or `NUL.` would successfully bypass the strict device name check but still resolve to the reserved device. I've updated the logic to use `.TrimEnd(' ', '.')` on the base name before checking it against the reserved device name list, and I've added test coverage for these edge cases.

File: src/Appium.Net/Appium/PathHelpers.cs [hunk 20]
```diff
@@ -0,0 +1,198 @@
...
+using System.Runtime.InteropServices;
+
+namespace OpenQA.Selenium.Appium
+{
+    internal static class PathHelpers
+    {
+        private static readonly System.Reflection.PropertyInfo LinkTargetProperty = 
+            typeof(FileSystemInfo).GetProperty("LinkTarget");
+
+        private static readonly bool IsWindows = RuntimeInformation.IsOSPlatform(OSPlatform.Windows);
```
[the remark this answers is not in the document]
Comment (Dor-bl) [reply src/Appium.Net/Appium/PathHelpers.cs:13]: Good catch. I've extracted the `HashSet` to a `private static readonly` field to avoid allocating it on every call.

File: src/Appium.Net/Appium/PathHelpers.cs [hunk 21]
```diff
@@ -0,0 +1,198 @@
...
+                string part = parts[i];
+
+                if (IsWindows)
+                {
+                    string baseName = Path.GetFileNameWithoutExtension(part).TrimEnd(' ', '.');
+                    if (reservedDeviceNames.Contains(baseName))
+                    {
+                        throw new ArgumentException("The file name contains a reserved device name.", nameof(fileName));
+                    }
+                }
```
[the remark this answers is not in the document]
Comment (Dor-bl) [reply src/Appium.Net/Appium/PathHelpers.cs:54]: Done. The local variable has been removed and the lookup now properly uses the static readonly set.
